### PR TITLE
Add OCI documentation

### PR DESCRIPTION
**Pull request by carabasdaniel**
Closes: https://github.com/open-policy-agent/opa/issues/4638

### CODE REVIEW

File: docs/content/management-bundles.md [hunk 1]
```diff
@@ -252,6 +252,49 @@ and number keys are converted to strings. Also, YAML !!binary tags are not
...
+
+To build and push a policy bundle to a remote OCI registry with the [OPA CLI](https://www.openpolicyagent.org/docs/edge/cli/) and [ORAS CLI](https://oras.land/cli/) you can  use the following commands:
+
+- `opa build <path_to_src>` will allow you to build a bundle tarball from your OPA policy and data files
+
+Now that we have the tarball we will need to provide a config manifest to the ORAS CLI and the tarball itself: 
+- `oras push <registry>/<org>/<repo>:<tag> --manifest-config <you_config_json>:application/vnd.oci.image.config.v1+json <the_tarball_obtained_from_opa_build>:application/vnd.oci.image.layer.v1.tar+gzip`
+
+Using an empty(`{}`) `manifest-config` json file should be sufficient to be able to push and allow the OCI downloader to use the remote policy image. 
+
```
Comment (srenatus) [inline docs/content/management-bundles.md:293]: For a complete documentation experience,

- Let's also add an example how you'd configure OPA to download that policy stored with an OCI registry
- Can I pin a version/tag/ref? 
- When will OPA redownload and activate an updated OCI bundle?

File: docs/content/management-bundles.md [hunk 2]
```diff
@@ -252,6 +252,182 @@ and number keys are converted to strings. Also, YAML !!binary tags are not
...
+OPA is able to interact with [OCI](https://opencontainers.org/) compatible registries to be able to download and use policies stored as containers.
+To configure OPA to use an OCI repository see the [service configuration section](../configuration/#services)
+
+**Structure**
+The bundle container is composed of 3 layers:
+- the manifest layer - contains the information about the tarball layer of the container(the digest, size, mediatype and annotations) and the config layer
+- the bundle tarball layer - the actual bundle tarball
+- the configuration layer - currently empty
+
+For OCI compatible registries an ***oci*** folder is created in the [persistence directory](../configuration/#miscellaneous). If this value is not set, because the OCI downloader plugin requires a storage path, the system's temporary folder location will be used instead.
```
Comment (srenatus) [inline docs/content/management-bundles.md:266]: Should the user care about this folder? Should it be backed-up or cleaned out regularly? 💭

File: docs/content/management-bundles.md [hunk 3]
```diff
@@ -252,6 +252,182 @@ and number keys are converted to strings. Also, YAML !!binary tags are not
...
+    - `application/vnd.oci.image.manifest.v1+json`
+    - `application/vnd.oci.image.config.v1+json`
+
+#### Building and Publishing Policy Containers
+
+There are multiple ways to build an image from a policy code base using different tools.
+
+##### Using OPA and ORAS CLIs
+
+To build and push a policy bundle to a remote OCI registry with the [OPA CLI](https://www.openpolicyagent.org/docs/edge/cli/) and [ORAS CLI](https://oras.land/cli/) you can  use the following commands:
```
Comment (srenatus) [inline docs/content/management-bundles.md:282]: [nit] This, too, can be a relative link `../cli/` for our CLI docs

File: docs/content/management-bundles.md [hunk 4]
```diff
@@ -252,6 +252,182 @@ and number keys are converted to strings. Also, YAML !!binary tags are not
...
+
+I'll prepare an empty config.json file that contains:
+```
+{}
+```
+
+To push the build image to an upstream registry we first need to login using:
+```
+ oras login ghcr.io
+```
```
Comment (srenatus) [inline docs/content/management-bundles.md:355]: [nit]
```suggestion
```bash
oras login ghcr.io
```
``

(also below)

**Review by srenatus — COMMENTED** [review]
Thanks for bearing with me! I've added a few more comments.

**Comment by srenatus** [comment]
This is also a good moment to drop `docs/devel/OCI.md`.

File: docs/content/management-bundles.md (hunk 2 above)
Comment (carabasdaniel) [reply docs/content/management-bundles.md]: Added a comment that this folder should be maintained by the user and it should be backed-up/cleaned periodically as this also acts as a cache for the OCI downloader

**Comment by carabasdaniel** [comment]
@srenatus I talked with the team about the policy CLI. This is an OSS tool similar to oras within it's own Github organization so we don't think this is really a vendor-specific tool as we wanted it to be independent.

**Comment by srenatus** [comment]
Thanks! I think what's left is pretty standard procedure -- let's remove the entire file, please 😃

**Comment by srenatus** [comment]
Hah, the failing test is #4748 🙃

**Comment by carabasdaniel** [comment]
This should close also https://github.com/open-policy-agent/opa/issues/4748

File: docs/content/management-bundles.md [hunk 5]
```diff
@@ -252,6 +252,189 @@ and number keys are converted to strings. Also, YAML !!binary tags are not
...
+    - `application/vnd.oci.image.manifest.v1+json`
+    - `application/vnd.oci.image.config.v1+json`
+
+#### Building and Publishing Policy Containers
+
+There are multiple ways to build an image from a policy code base using different tools.
+
+#### Using policy CLI
+
+The [policy CLI](https://www.openpolicyregistry.io/docs/cli/download) tool can be easily used to build and push a policy bundle to a remote OCI registry using just two simple commands:
```
Comment (srenatus) [inline docs/content/management-bundles.md:282]: Let's keep OPR and the policy CLI out of it, please. It's vendor-provided tooling, we don't want to link to that. Please refer to the OPA docs in the OPR docs, but not this way around.

File: docs/content/management-bundles.md [hunk 6]
```diff
@@ -252,6 +252,189 @@ and number keys are converted to strings. Also, YAML !!binary tags are not
 supported.
 {{< /info >}}
 
+### Bundle Container Format
```
Comment (srenatus) [inline docs/content/management-bundles.md:255]: This section no longer fits in. For "Bundle Container Format", it made sense to have it follow "Bundle File Format", but what we really need is a separate OCI section at the bottom the bundle docs.

💭 Maybe it would make more sense to have an "OCI Registry" section next to the other "Implementations", like, following Nginx.

File: docs/content/management-bundles.md [hunk 7]
```diff
@@ -252,6 +252,189 @@ and number keys are converted to strings. Also, YAML !!binary tags are not
...
+#### Using policy CLI
+
+The [policy CLI](https://www.openpolicyregistry.io/docs/cli/download) tool can be easily used to build and push a policy bundle to a remote OCI registry using just two simple commands:
+- `policy build <path_to_src> -t <org>/<repo>:<tag>`
+- `policy push <registry>/<org>/<repo>:<tag>`
+A full tutorial is available [here](https://www.openpolicyregistry.io/docs/tutorial)
+
+##### Using OPA and ORAS CLIs
+
+To build and push a policy bundle to a remote OCI registry with the [OPA CLI](https://www.openpolicyagent.org/docs/edge/cli/) and [ORAS CLI](../cli/) you can  use the following commands:
```
Comment (srenatus) [inline docs/content/management-bundles.md:289]: The other relative link please -- OPA's CLI, not ORAS 😉

File: docs/content/management-bundles.md [hunk 8]
```diff
@@ -252,6 +252,189 @@ and number keys are converted to strings. Also, YAML !!binary tags are not
...
+OPA is able to interact with [OCI](https://opencontainers.org/) compatible registries to be able to download and use policies stored as containers.
+To configure OPA to use an OCI repository see the [service configuration section](../configuration/#services)
+
+**Structure**
+The bundle container is composed of 3 layers:
+- the manifest layer - contains the information about the tarball layer of the container(the digest, size, mediatype and annotations) and the config layer
+- the bundle tarball layer - the actual bundle tarball
+- the configuration layer - currently empty
+
+For OCI compatible registries an ***oci*** folder is created in the [persistence directory](../configuration/#miscellaneous). If this value is not set, because the OCI downloader plugin requires a storage path, the system's temporary folder location will be used instead. This folder should be maintained by the user. We recommend backing-up or cleaning up this folder periodically as this acts as a local cache for the OCI downloader.
```
Comment (srenatus) [inline docs/content/management-bundles.md:266]: Can I have bundles pulled from multiple registries? I suppose so. But one, shared, `oci` folder is still OK?

**Review by srenatus — COMMENTED** [review]
Some questions/notes/requests. Thanks for bearing with me

File: docs/content/management-bundles.md (hunk 8 above)
Comment (carabasdaniel) [reply docs/content/management-bundles.md]: Yes, it should still be ok

**Comment by srenatus** [comment]
> @srenatus I talked with the team about the policy CLI. This is an OSS tool similar to oras within it's own Github organization so we don't think this is really a vendor-specific tool as we wanted it to be independent.

The `policy` CLI doesn't compare to ORAS, I think. ORAS is not specific in any way: it's a generic OCI tool, like curl is for HTTP, as far as I am concerned. On the one hand, that link goes to the OPCR docs, not the github repo for `policy`. The github repo also states that it's "for OPCR" (which it talks to by default) which is not an OSS tool.

Both OPCR and the `policy` tool (and Aserto, for that matter) are on the ecosystem page already. I propose we keep it at that. Perhaps it would make sense to add the word "OCI" somewhere there so it's more obvious what it's about?

File: test/e2e/oci/oci_test.go [hunk 9]
```diff
@@ -109,6 +109,7 @@ func TestEnablePrintStatementsForBundles(t *testing.T) {
 		if !strings.Contains(buf.String(), expContains) {
 			t.Fatalf("expected logs to contain %q but got: %v", expContains, buf.String())
 		}
+		time.Sleep(3 * time.Second) // wait a couple of seconds for the second trigger to kick in
```
Comment (JAORMX) [inline test/e2e/oci/oci_test.go:112]: Seems to me like this should be included in a separate PR.

**Review by JAORMX — CHANGES_REQUESTED** [review]

File: test/e2e/oci/oci_test.go (hunk 9 above)
Comment (srenatus) [reply test/e2e/oci/oci_test.go:112]: :+1: spinning this out into an extra PR is a good idea!

Comment (carabasdaniel) [reply test/e2e/oci/oci_test.go:112]: 👍

Comment (carabasdaniel) [reply test/e2e/oci/oci_test.go]: Extracted into https://github.com/open-policy-agent/opa/pull/4806

**Comment by ogazitt** [comment]
> > @srenatus I talked with the team about the policy CLI. This is an OSS tool similar to oras within it's own Github organization so we don't think this is really a vendor-specific tool as we wanted it to be independent.
> 
> The `policy` CLI doesn't compare to ORAS, I think. ORAS is not specific in any way: it's a generic OCI tool, like curl is for HTTP, as far as I am concerned. On the one hand, that link goes to the OPCR docs, not the github repo for `policy`. The github repo also states that it's "for OPCR" (which it talks to by default) which is not an OSS tool.
> 
> Both OPCR and the `policy` tool (and Aserto, for that matter) are on the ecosystem page already. I propose we keep it at that. Perhaps it would make sense to add the word "OCI" somewhere there so it's more obvious what it's about?

Hi @srenatus, OPCR is meant to be a reference implementation of a "policy registry" - it is single-user, free to use, and has no vendor angle (or upsell). And the `policy` CLI is built to work with any OCI-compatible registry, and AFAIK the github repo doesn't contain any reference to OPCR in particular. 

Stepping back, the entire point of the `policy` CLI is to make it easy to build OPA policies into OCI images, which hopefully we agree is a good thing. It was built as an OSS tool, in service to the broader OPA community, to support any use-case involving OPA (gatekeeper, conftest, etc). It's in a vendor-neutral GH organization (opcr-io). We would gladly house it under the OPA project if that would remove the concern.

**Comment by srenatus** [comment]
👋 hello @ogazitt welcome to this discussion.

> OPCR is meant to be a reference implementation of a "policy registry" - it is single-user, free to use, and has no vendor angle (or upsell). And the policy CLI is built to work with any OCI-compatible registry, and AFAIK the github repo doesn't contain any reference to OPCR in particular.

OPCR is the default server that `policy` uses, and the github repo says "Policy CLI for the Open Policy Registry (opcr.io)".

But really, I believe OPCR doesn't need this extra mention. Besides being on the ecosystem page _already_, we have no evidence that our users care for it, and I don't think we should push them into using the "reference implementation". Frankly, since OPCR isn't open source, I'm not sure I see the "reference" angle here -- we don't learn anything from using OPCR that we wouldn't learn from using AWS' or GitHub's OCI registries.

Also, the docs here show that there's not much that it does that a shell script using oras or any other standard OCI tool wouldn't be able to accomplish. If there was much effort, we'd need to provide that in `opa push`, or some similar kind of out-of-the-box experience 🤔

**Comment by ogazitt** [comment]
> 👋 hello @ogazitt welcome to this discussion.
> 
> > OPCR is meant to be a reference implementation of a "policy registry" - it is single-user, free to use, and has no vendor angle (or upsell). And the policy CLI is built to work with any OCI-compatible registry, and AFAIK the github repo doesn't contain any reference to OPCR in particular.
> 
> OPCR is the default server that `policy` uses, and the github repo says "Policy CLI for the Open Policy Registry (opcr.io)".
> 
> But really, I believe OPCR doesn't need this extra mention. Besides being on the ecosystem page _already_, we have no evidence that our users care for it, and I don't think we should push them into using the "reference implementation". Frankly, since OPCR isn't open source, I'm not sure I see the "reference" angle here -- we don't learn anything from using OPCR that we wouldn't learn from using AWS' or GitHub's OCI registries.
> 
> Also, the docs here show that there's not much that it does that a shell script using oras or any other standard OCI tool wouldn't be able to accomplish. If there was much effort, we'd need to provide that in `opa push`, or some similar kind of out-of-the-box experience 🤔

Ah, I see the description of the [policy](https://github.com/opcr-io/policy) repo does contain a mention of Open Policy Registry, and I can see why that would be a concern, because `policy` was designed to work with any oras-compatible container registry. I assuming revising the description to "CLI for building OPA policies into OCI images" would alleviate this concern?

**Comment by philipaconrad** [comment]
After taking a look at the docs text, I can see what @srenatus is getting at. `policy` isn't a particularly necessary part of that documentation addition. I think if we axe the `policy`-specific section, that should be sufficient for getting this PR finally shipped, because the rest looks pretty good.

Related: I think that if we find users often want to be building OCI images for OPA, that ought to become part of OPA itself, maybe as part of what a sub-command on the CLI (like `opa push`?) handles. :thinking:

**Comment by carabasdaniel** [comment]
Hi @philipaconrad @srenatus, 
I've updated the PR but the failing test does not seem related to these changes, can you please let me know if that is the case ?

**Comment by philipaconrad** [comment]
@carabasdaniel You're correct. The test failure is unrelated to these changes. Our MacOS tests have been a bit flaky recently. We'll re-run the failing CI tests, and this PR should go green again.

File: docs/content/management-bundles.md (hunk 5 above)
Comment (philipaconrad) [reply docs/content/management-bundles.md]: The PR was updated recently to address this concern. I'm marking this as resolved. :+1:

**Review by philipaconrad — APPROVED** [review]
These changes look ready for merge! :+1:
